load extensions through pythons entry-points machinery

tjfulle opened the pull request:
## What is it?

This PR adds the ability to load spack extensions through `importlib.metadata.entry_points` in addition to the regular configuration variable.

## Why?

This allows one to `pip install my-extension` and have spack find it without having to provide additional configuration.

## Notes

- See initial discussion at https://spackpm.slack.com/archives/C5W7NKZJT/p1706560654393159
- Documentation and tests have not been provided, but can if this PR is deemed useful and/or of interest.

## Usage

In your package's `pyproject.toml` add the following:

```toml
[project.entry-points."spack.extensions"]
"ext-name" = "module-name:function-that-returns-list-of-extension-paths"

[project.entry-points."spack.config"]
"ext-name" = "module-name:function-that-returns-config-scope-path"
```

## Example

`pyproject.toml`

```toml
[project]
name = "baz"
description = "baz"
version = "0.1.0"
requires-python = ">=3.9"

[project.entry-points."spack.extensions"]
baz = "baz:get_spack_extension_paths"

[project.entry-points."spack.config"]
baz = "baz:get_spack_config_dir"
```

`baz/__init__.py`:

```python
import importlib.resources

def get_spack_extension_paths() -> list[str]:
    dirname = importlib.resources.files("baz").joinpath("spack-baz")
    if dirname.exists():  # type: ignore
        return [str(dirname)]
    return []

def get_spack_config_dir() -> str:
    dirname = importlib.resources.files("baz").joinpath("spack-config")
    if dirname.exists():  # type: ignore
        return str(dirname)
    return []
```

tjfulle committed:
load extensions through pythons entry-points machinery

tjfulle committed:
read entry points from different python versions correctly

tjfulle commented:
@psakievich

psakievich commented:
This looks valuable to me.  Personally, I'm okay with this feature being conditionally available if `SPACK_PYTHON>=3.8` but I think we need to make sure the imports are conditional on python version so we can support older python versions that spack still supports.

tjfulle committed:
try: import importlib.metadata to work with python < 3.8

tjfulle commented:
The last commit hides `importlib.metadata` (which is available in python >= 3.8) so that this PR is compatible with older python.  Older python will not be able to use the entry points as a means to inject user configurations into spack.

tjfulle committed:
black/flake8

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -763,6 +768,22 @@ def _add_platform_scope(
+def _add_configuration_paths_from_entry_points(configuration_paths):
+    if importlib_metadata is None:
+        return
+    try:
+        entry_points = importlib_metadata.entry_points(group="spack.config")
+    except TypeError:
+        entry_points = importlib_metadata.entry_points().get("spack.config")
```
I'm not super familiar with entry points. Can you explain when we'll get the "try" succeeding, and when we fall back on the exceptional case?

tjfulle replied:
@alalazo - the exceptional case is for Python 3.8, the default case is for 3.9 and newer.

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -763,6 +768,22 @@ def _add_platform_scope(
     cfg.push_scope(scope_type(plat_name, plat_path))
 
 
+def _add_configuration_paths_from_entry_points(configuration_paths):
```
(minor) Here I would just define:
```suggestion
def paths_from_entry_points() -> List[str]:
```
and return the list of paths from entry points

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -815,6 +836,8 @@ def create() -> Configuration:
     # No site-level configs should be checked into spack by default.
     configuration_paths.append(("site", os.path.join(spack.paths.etc_path)))
 
+    _add_configuration_paths_from_entry_points(configuration_paths)
```
(minor) While here I would:
```suggestion
    configuration_paths.extend(paths_from_entry_points())
```

alalazo commented on lib/spack/spack/extensions.py:
```diff
@@ -132,10 +137,31 @@ def load_extension(name: str) -> str:
+
+    """
+    extension_paths = []
+    if importlib_metadata is None:
+        return extension_paths
+    try:
+        entry_points = importlib_metadata.entry_points(group="spack.extensions")
+    except TypeError:
+        entry_points = importlib_metadata.entry_points().get("spack.extensions")
```
This code and the import above are duplicated. They can probably be factored in a separate function and used both here and in `spack.config`.

alalazo reviewed (commented):
I like the idea, it might even blend well with #40773 and the other idea of being able to turn Spack extensions into Spack packages.

tjfulle committed:
addressing feedback from @alalazo and adding documentation

tjfulle committed:
fix typo in docs

tjfulle committed:
fix typo in docs

tjfulle committed:
fix typo in docs

tjfulle committed:
fix some styling

tjfulle committed:
fixing up stying issues

tjfulle commented:
Thanks for the feedback @alalazo !

I addressed your comments and added documentation.  I'll likely need some help testing.  A test would require creating a python venv, installing a dummy package into it, then testing whether spack can read the entry points.  Or, that could all be mocked.  Not sure which is best...

tjfulle committed:
adding a reference to the hypothesis implementation

tjfulle committed:
only call entry point hook if callable

tjfulle committed:
get past some mypy errors

tjfulle committed:
fix read the docs error

tjfulle commented:
@alalazo  and @psakievich : bump

psakievich reviewed (dismissed):
This looks good to me. @alalazo thoughts on the changes?

alalazo commented:
@psakievich I like the idea, but this PR needs to pass CI and add some unit tests before being merged. Also, we need to state in the docs what are the additional requests we make for people who want to use this feature (I guess either Python > 3.8 or `setuptools` installed).

tjfulle committed:
adding entry points test

tjfulle committed:
mock out entry point test

tjfulle committed:
remove debug print

tjfulle committed:
adding some novermin to pass ci testing

tjfulle committed:
add more tests

tjfulle committed:
update docs to note that python 3.8 or newer is required for entry_points

tjfulle committed:
fixing style issue

tjfulle committed:
hopefully this will pass with windows pathseps...

tjfulle committed:
more extension functionality testing

tjfulle committed:
remove debug prints

tjfulle commented:
thanks for the feedback @alalazo

- documents have been updated with a not about requiring `python>=3.8` 
- 2 unit tests give nearly 100% coverage of new code added, with the exception of some defensive code for `python 3.7` (which doesn't get hit by the ci runner)

alalazo commented:
@tjfulle Thanks! I'll review either today or beginning of next week.

tjfulle commented:
@alalazo said:

> I like the idea, it might even blend well with https://github.com/spack/spack/pull/40773 and the other idea of being able to turn Spack extensions into Spack packages.

Incidentally, this is essentially what we are doing internally in my group at Sandia: `python -m pip install <my group's extensions>` and then Spack "just works" with our extensions, repos, configurations, etc.  If the extension were packaged as a `PythonPackage`, then just replace `python -m pip install` with `spack install`.  No further work needed and the Spack extensions are ready to roll.

alalazo commented on lib/spack/spack/test/entry_points.py:
```diff
@@ -0,0 +1,111 @@
+    return entry_points
+
+
+def patch_entry_points(tmpdir, monkeypatch):
```
(minor) You can turn this into a fixture, maybe use `tmp_path` instead of `tmpdir` (deprecated):
```suggestion
@pytest.fixture()
def mock_entry_points(tmp_path, monkeypatch):
```

alalazo commented on lib/spack/spack/test/entry_points.py:
```diff
@@ -0,0 +1,111 @@
+
+
+@pytest.mark.skipif(sys.version_info[:2] < (3, 8), reason="Python>=3.8 required")
+def test_spack_entry_point_config(monkeypatch, tmpdir):
+    """Test config scope entry point"""
+    patch_entry_points(tmpdir, monkeypatch)
```
Then this will look like (edited in web UI, so beware of typos):
```suggestion
def test_spack_entry_point_config(tmp_path, mock_entry_points):
    """Test config scope entry point"""    
```

alalazo commented on lib/spack/spack/test/entry_points.py:
```diff
@@ -0,0 +1,111 @@
+    patch_entry_points(tmpdir, monkeypatch)
+    config_paths = dict(spack.config.config_paths_from_entry_points())
+    config_path = config_paths.get("plugin-mypackage_config")
+    my_config_path = os.path.join(tmpdir, "spack", "etc")
```
Since `tmp_path` is a `pathlib.Path` object:
```suggestion
    my_config_path = tmp_path / "spack" / "etc"
```

alalazo commented on lib/spack/spack/test/entry_points.py:
```diff
@@ -0,0 +1,111 @@
+
+
+@pytest.mark.skipif(sys.version_info[:2] < (3, 8), reason="Python>=3.8 required")
+def test_spack_entry_point_extension(monkeypatch, tmpdir):
```
Similarly here

alalazo reviewed (changes requested):
Thanks! This LGTM, I just left a couple of minor comments for tests. Let me know if you agree with them. In the meanwhile I'll test drive the PR, but I expect to merge it shortly.

alalazo commented:
@tjfulle Would you agree putting a note in the docs saying that the entry-point support is still experimental and we might change / improve the layout anytime? I'd like to merge this one soon, but I'm wondering if we can simplify the required layout:
```
my-package/
├── lib
│   ├── my_package
│   │   └── __init__.py
│   └── spack-scripting/  # the spack extensions
└── pyproject.toml
```

tjfulle committed:
updating entry points test to use fixtures and tmp_path.  also allowing extension paths to be Path objects

tjfulle commented:
@alalazo - thanks for the feedback!  I updated the test and the library function to accept `pathlib.Path` objects.

tjfulle commented:
@alalazo said:

> Would you agree putting a note in the docs saying that the entry-point support is still experimental and we might change / improve the layout anytime? I'd like to merge this one soon, but I'm wondering if we can simplify the required layout:

The layout in the documentation is the "`src`" layout recommended by the [python packaging authority](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/).  I would recommend leaving it.  Furthermore, I don't think Spack should make a recommendation (outside of what [pypa](https://packaging.python.org/en/latest/) has already recommended) on a python package's layout.  It should just promise that if a package registers entry point hooks the meet the contract, spack will load them.  The user can use whatever layout they choose.

tjfulle committed:
fix typo in docs, should be src not lib for python package layout

tjfulle commented:
just noticed now that the package layout in the examples used `lib` instead of `src`.  It's fixed now.

alalazo commented:
> The layout in the documentation is the "src" layout recommended by the [python packaging authority](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/). [ .. ]

I agree with what you say, what bother me is more on Spack side - the necessity to have a `spack-xxx` subfolder within another package :thinking:[^1] In any case not a deal-breaker. We can try to simplify that later in case.

[^1]: This is mostly due to the fact that we didn't design extension with entry points in mind years ago.

alalazo reviewed (approved):
Thanks!

haampie commented:
The other minor point about this PR is that it triggers a system scan on every Spack invocation:

```
$ strace -e file -f python3 -c 'from importlib import metadata; metadata.entry_points()' |& grep entry_points
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/packaging-23.1.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/decorator-5.1.1.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/cycler-0.11.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/numpy-1.26.3.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/jmespath-1.0.1.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/kiwisolver-1.4.5.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/isort-5.12.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/fonttools-4.39.4.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/s3transfer-0.6.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/boto3-1.26.26.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/pickleshare-0.7.5.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/click-8.1.7.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/executing-1.2.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/platformdirs-3.10.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/mypy_extensions-1.0.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/vermin-1.6.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = 3
...
```

not as bad as Spack's own package cache invalidation, but still.

tjfulle commented:
@haampie said:

> is there any mechanism in entry_points.txt that ensures compatibility with Spack version and Python interpreter?

`importlib.metadata.entry_points` was introduced in Python 3.8 as a standard way of extending python packages.  It ships with the Python interpreter's standard library, so I am not sure what you mean by ensure compatibility with the interpreter.

@haampie said: 

> Can it trigger arbitrary function calls?

`importlib.metada.entry_points` won't trigger arbitrary function calls.  It will return an entry point that Spack loads.  The loading process will trigger whatever function calls are in the entry point.
